docs/features/AnoncredsControllerMigration.md

```diff
@@ -0,0 +1,668 @@
+
+#### Get registries
+
+- Change from **GET /revocation/registries/created** to **GET /anoncreds/revocation/registries**
+- No payload changes
+
+#### Get registry
+
+- Changes from **GET /revocation/registry/{rev_reg_id}** to **GET /anoncreds/revocation/registry/{rev_reg_id}**
+- No payload changes
+
+#### Fix reocation state
+
+- Changes from **POST /revocation/registry/{rev_reg_id}/fix-revocation-entry-state** to **POST /anoncreds/revocation/registry/{rev_reg_id}/fix-revocation-state**
+- No payload changes
+
+#### Get number of issued credentials
+
+- Changes from **GET /revocation/registry/{rev_reg_id}/issued** to **GET /anoncreds/revocation/registry/{rev_reg_id}/issued**
+- No payload changes
+
+#### Get credential details
+
+- Changes from **GET /revocation/registry/{rev_reg_id}/issued/details** to **GET /anoncreds/revocation/registry/{rev_reg_id}/issued/details**
+- No payload changes
+
+#### Get revoked credential details
+
+- Changes from **GET /revocation/registry/{rev_reg_id}/issued/indy_recs** to **GET /anoncreds/revocation/registry/{rev_reg_id}/issued/indy_recs**
+- No payload changes
+
+#### Set state manually
+
+- Changes from **PATCH /revocation/registry/{rev_reg_id}/set-state** to **PATCH /anoncreds/revocation/registry/{rev_reg_id}/set-state**
+- No payload changes
+
+#### Upload tails file
+
+- Changes from **PUT /revocation/registry/{rev_reg_id}/tails-file** to **PUT /anoncreds/registry/{rev_reg_id}/tails-file**
+- No payload changes
+
+#### Download tails file
+
+- Changes from **GET /revocation/registry/{rev_reg_id}/tails-file** to **GET /anoncreds/revocation/registry/{rev_reg_id}/tails-file**
+- No payload changes
+
+#### Revoke a credential
+
+- Changes from **POST /revocation/revoke** to **POST /anoncreds/revocation/revoke**
+- Change payload and parameters from
+
+#### Clear pending revocations
+
+- **POST /revocation/clear-pending-revocations** has been removed.
+
+#### Delete tails file
+
+- Endpoint **DELETE /revocation/delete-tails-server** has been removed
+
+#### Update tails file
+
+- Endpoint **PATCH /revocation/registry/{rev_reg_id}** has been removed
+
+#### Additional Endpoints
+
+- **PUT /anoncreds/registry/{rev_reg_id}/active** is available to set the active registry
+
+## Step 2 - Trigger the upgrade via the upgrade endpoint
+
+The upgrade endpoint is at **POST /anoncreds/wallet/upgrade**.
+
+You need to be careful doing this, as there is no way to downgrade the wallet. Once the wallet is upgraded it will not be able to interact with agents that have not been upgraded. It is recommended to test the upgrade in a development environment before upgrading a production wallet.
+
+Params: `wallet_name` is the name of the wallet to upgrade. Used to prevent accidental upgrades.
+
+The behavior for a base wallet (standalone) or admin wallet in multitenant mode is different from the behavior of a subwallet (or tenant) in multitenancy mode.
+
+### Base wallet (standalone) or admin wallet in multitenant mode:
+
```

Comment:
Assuming they are **NOT** upgrading Askar (which is a different story):

* Why is the aca-py update needed as part of the process?  E.g. shouldn’t that be a separate topic?
  * E.g., the phrasing is more like — you must be running ACA-Py 0.x.0 at least.
* For a data change upgrade like this one, are steps 2 and 3 required? (e.g. no database schema change, just existing records are updated).
* Is there any way to use a semaphore to support starting up multiple controllers on step 4 such that all try to execute the upgrade, one carries it out, all wait until it is done, and all continue after?
  * Or at least — one completes the actual upgrade, the rest execute the “null op” upgrade since it is already. complete.
* For a non-“Indy Wallet to Askar” upgrade, why the restart at the end?

Reply:
- Yes i think the phasing should be changed. Basically they just need to have the version with the upgrade endpoint.
- There is schema changes with the upgrade. See https://github.com/hyperledger/aries-cloudagent-python/issues/2384.
- There is a semaphore and this should work. There is still a question about if this can be done with a cache. Right now the semaphore is a database record that needs to be queried in the middleware. Which isn't ideal.
- The restart is only needed because the agent expects to have the correct `wallet-type` in the startup configuration. I had been working on a way for this not to be required but it was starting to make the PR scope get out of control. I also wasn't sure the best way to go about it. Not having to restart would be ideal.